packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -70,6 +70,20 @@ export class ModuleGraph {
     return this.fileToModulesMap.get(file)
   }
 
+  addWatchModuleToFile(file: string, id: string): void {
+    file = normalizePath(file)
+    let fileMappedModules = this.fileToModulesMap.get(file)
+    if (!fileMappedModules) {
+      fileMappedModules = new Set()
+      this.fileToModulesMap.set(file, fileMappedModules)
+    }
+
+    const moduleNode = this.idToModuleMap.get(id)
+    if (moduleNode) {
+      fileMappedModules.add(moduleNode)
+    }
+  }
```

Comment:
@mattcompiles what do you think if replace `addWatchModuleToFile` with

```js
  ensureModulesByFile(file: string): void {
    file = normalizePath(file)
    let fileMappedModules = this.fileToModulesMap.get(file)
    if (!fileMappedModules) {
      fileMappedModules = new Set()
      this.fileToModulesMap.set(file, fileMappedModules)
    }
    return fileMappedModules
  }
```

So we can then use it in `transformRequest.ts`.

```js
      transformResult.watchFiles.forEach((file) => {
        moduleGraph.ensureModulesByFile(file).add(mod)
      })
```

If we end up going this way, I think it is better to avoid the ( mod -> mod.id -> mod ) currently present in the PR. `ensureModulesByFile` could also be used in another PR to refactor and simplify a few function implementations in the moduleGraph.

Something I forgot yesterday, last time that we discussed https://github.com/vitejs/vite/issues/3216 we saw that there was a lot of support (a lot of +1 reactions) but there wasn't much in concrete examples of where this feature would help. If you have a use case for this feature, it would be great if you could add a comment in the issue describing it so we leave it as reference.

Reply:
Happy for any suggested strategies by the core team. I'll wait for a final opinion before updating the code though if that's ok. 

[Added my use-case in the issue.](https://github.com/vitejs/vite/issues/3216#issuecomment-890752320)